crates/core/src/operations/add_column.rs

```diff
@@ -0,0 +1,125 @@
+            log_store,
+            fields: None,
+            commit_properties: CommitProperties::default(),
+        }
+    }
+
+    /// Specify the fields to be added
+    pub fn with_fields(mut self, fields: Vec<StructField>) -> Self {
+        self.fields = Some(fields);
+        self
+    }
+    /// Additional metadata to be added to commit info
+    pub fn with_commit_properties(mut self, commit_properties: CommitProperties) -> Self {
+        self.commit_properties = commit_properties;
+        self
+    }
+}
+
+impl std::future::IntoFuture for AddColumnBuilder {
+    type Output = DeltaResult<DeltaTable>;
+
+    type IntoFuture = BoxFuture<'static, Self::Output>;
+
+    fn into_future(self) -> Self::IntoFuture {
+        let this = self;
+
+        Box::pin(async move {
+            let mut metadata = this.snapshot.metadata().clone();
+            let fields = match this.fields {
+                Some(v) => v,
+                None => return Err(DeltaTableError::Generic("No fields provided".to_string())),
+            };
+
+            let table_schema = this.snapshot.schema();
+            let new_table_schema =
+                merge_struct(table_schema, &fields.clone().into_iter().collect())?;
+
+            // TODO(ion): Think of a way how we can simply this checking through the API or centralize some checks.
+            let contains_timestampntz = PROTOCOL.contains_timestampntz(&fields);
+            let protocol = this.snapshot.protocol();
+
+            let maybe_new_protocol = if contains_timestampntz {
+                if protocol.min_reader_version == 3 && protocol.min_writer_version == 7 {
+                    let mut new_protocol = protocol.clone();
+                    new_protocol = new_protocol
+                        .with_reader_features(vec![ReaderFeatures::TimestampWithoutTimezone]);
+                    new_protocol = new_protocol
+                        .with_writer_features(vec![WriterFeatures::TimestampWithoutTimezone]);
+                    Some(new_protocol)
+                } else {
+                    let new_protocol = Protocol {
+                        min_reader_version: 3,
+                        min_writer_version: 7,
+                        writer_features: Some(hashset! {WriterFeatures::TimestampWithoutTimezone}),
+                        reader_features: Some(hashset! {ReaderFeatures::TimestampWithoutTimezone}),
+                    };
+                    // Convert existing properties to features since we advance the protocol to v3,7
+                    Some(convert_properties_to_features(
+                        new_protocol,
+                        &metadata.configuration,
+                    ))
+                }
+            } else {
+                None
+            };
+
+            let operation = DeltaOperation::AddColumn {
+                fields: fields.into_iter().collect_vec(),
+            };
+
+            metadata.schema_string = serde_json::to_string(&new_table_schema)?;
```

Comment:
Just thinking out loud here, not feedback on this pull request, but implementing TryInto<String> for StructType would probably be handle for us